### PR TITLE
DAOS-15125 build: Accept rc versions when checking for go support.

### PR DESCRIPTION
**Pull request by ashleypittman**
- **DAOS-15125 build: Accept rc versions when checking for go support.**

### CODE REVIEW

**Comment by ashleypittman** [comment]
This is to fix an issue with functional on leap testing but I can't find the right commit tags to enable that branch.

**Comment by ashleypittman** [comment]
It might be "Skip-func-test-leap15: False", I'll try with that if the current testing doesn't catch it.

File: site_scons/site_tools/go_builder.py [hunk 1]
```diff
@@ -49,6 +49,28 @@ def _scan_go_file(node, env, _path):
...
+    go_version = output.split(' ')[2].replace('go', '')
+    if '-' in go_version:
+        go_version = go_version.split('-')[0]
+    parts = []
+    for part in go_version.split("."):
+        if 'rc' in part:
+            parts.append(part.split("rc", maxsplit=1)[0])
+        else:
+            parts.append(part)
+    return ".".join(parts)
```
Comment (daltonbohning) [inline site_scons/site_tools/go_builder.py:63]: I know you're not usually a fan of regex, but...
```
def get_go_version(output):
    # Capture only the number, ignoring optional rc
    one_part = '([0-9]+)(?:rc[0-9]+)?'
    return '.'.join(re.findall(fr'(?:go)?{one_part}\.{one_part}\.{one_part}', output)[0])
```

**Review by daltonbohning — APPROVED** [review]

File: site_scons/site_tools/go_builder.py (hunk 1 above)
Comment (ashleypittman) [reply site_scons/site_tools/go_builder.py:63]: I'll try it out once the current test has completed and I can add a test for the actual output that we're actually parsing here.

**Comment by ashleypittman** [comment]
> > It might be "Skip-func-test-leap15: False", I'll try with that if the current testing doesn't catch it.
> 
> Yeah, I think so https://daosio.atlassian.net/wiki/spaces/CI/pages/1389759093/Commit+Pragmas
> 
> > `Skip-func-test-<distro>: true`

Ah, I was going by the text in the Explanation without reading it properly.

**Comment by daltonbohning** [comment]
> According to the logs the go version that we're using in CI is... "1.22.0" so I'm no longer able to reproduce or therefore test this, I think we should land it as-is.

I think the unit tests (manual) you added for the function most likely cover the rc

**Comment by ashleypittman** [comment]
> > According to the logs the go version that we're using in CI is... "1.22.0" so I'm no longer able to reproduce or therefore test this, I think we should land it as-is.
> 
> I think the unit tests (manual) you added for the function most likely cover the rc

the last push I made was to have it report the non-corrected version on stdout.  Let me check if it's doing that.

**Comment by ashleypittman** [comment]
> > According to the logs the go version that we're using in CI is... "1.22.0" so I'm no longer able to reproduce or therefore test this, I think we should land it as-is.
> 
> I think the unit tests (manual) you added for the function most likely cover the rc

It's "go version go1.22.0 linux/amd64".

File: site_scons/site_tools/go_builder.py [hunk 2]
```diff
@@ -49,6 +49,20 @@ def _scan_go_file(node, env, _path):
...
+    one_part = '([0-9]+)(?:rc[0-9]+)?'
+    return '.'.join(re.findall(fr'(?:go)?{one_part}\.{one_part}\.{one_part}', output)[0])
+
+
+def test_go():
+    """Quick unit test"""
+    assert "1.20.10" == get_go_version("go version go1.20.10 linux/amd64")
+    assert "1.2.3" == get_go_version("go version go1.2.3 Linux/amd64")
+    assert "1.20.10" == get_go_version("go version go1.20.10-daos linux/amd64")
+    assert "1.20.10" == get_go_version("go version go1.20rc2.10 linux/amd64")
```
Comment (brianjmurrell) [inline site_scons/site_tools/go_builder.py:63]: Are `go1.20rc2.10` and `go1.20.10` actually the same thing given that this function is returning them as equivalent?  If not will this ever be a problem with regard to minimum version matching?

Comment (ashleypittman) [reply site_scons/site_tools/go_builder.py:63]: This check appears to be there to enforce a minimum go compiler version so that we can use language features without further checking for them in the code, therefore I would expect those two to be the same.

Comment (ashleypittman) [reply site_scons/site_tools/go_builder.py:63]: I would also like to know that, in fact if you read the ticket you can see I've been asking exactly the same question.

The ticket relates to the daos_build test on Leap, it appears that's whats happening is that daos is built into an RPM and is happy with the scons version at that point, then the test tries to build daos over dfuse for testing and at that point the build of daos is failing with a go version that it can not parse.  I do not know the precise version of go that is being used for the test as the build did not log it (that's one change in this PR) and also the test is no longer failing, presumably as there's been another update to a non-rc version.

I neither know why Leap15 build and test were using different go versions, which the exact version was that was failing or what changed to enable the test to work.

I assume that it was using a rc build of 1.20 but that's purely an assumption, I just don't have more to go on despite looking for the information.

The test is passing again now, this PR should allow it to continue to work if a rc version is encountered again, plus as I say it'll log the actual version string this time so if this ever becomes an issue again we have more to work with.

Comment (brianjmurrell) [reply site_scons/site_tools/go_builder.py:63]: Should we hold off on landing this then until we have a test case in hand that we can further understand the situation and answer some of the open questions?

My concern is just landing something that provides a false report on the **go** version being used in cases where one of these mysterious _rc_ releases shows up.

**Review by jolivier23 — APPROVED** [review]
Added Mike to the list since he is more familiar with go and what the version might look like.

**Review by mjmac — CHANGES_REQUESTED** [review]
This won't work with versions that are hyphenated, e.g.:
```
go version go1.22-20240109-RC01 cl/597041403 +dcbe772469 X:fieldtrack,boringcrypto linux/amd64
```

I'll propose a solution in a bit.

File: site_scons/site_tools/go_builder.py [hunk 3]
```diff
@@ -49,6 +49,20 @@ def _scan_go_file(node, env, _path):
     return includes
 
 
+def get_go_version(output):
+    """Capture only the number, ignoring optional rc"""
+    one_part = '([0-9]+)(?:rc[0-9]+)?'
+    return '.'.join(re.findall(fr'(?:go)?{one_part}\.{one_part}\.{one_part}', output)[0])
+
+
+def test_go():
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply site_scons/site_tools/go_builder.py:58]: Manually, but it's a standard python test so either `pytest <file>` or `python -m pytest <file>` should run it.

File: site_scons/site_tools/go_builder.py [hunk 4]
```diff
@@ -49,6 +49,20 @@ def _scan_go_file(node, env, _path):
     return includes
 
 
+def get_go_version(output):
+    """Capture only the number, ignoring optional rc"""
+    one_part = '([0-9]+)(?:rc[0-9]+)?'
+    return '.'.join(re.findall(fr'(?:go)?{one_part}\.{one_part}\.{one_part}', output)[0])
```
Comment (mjmac) [inline site_scons/site_tools/go_builder.py:55]:
```suggestion
"""Capture only the version after 'go'"""
ver_re = re.compile(r'go([0-9\.]+)')
groups = ver_re.findall(output)
if not groups or len(groups) == 0:
    return ""
return groups[0]
```

File: site_scons/site_tools/go_builder.py (hunk 2 above)
Comment (mjmac) [inline site_scons/site_tools/go_builder.py:63]: This is a pretty bizarre version, but I don't think it makes sense to contort ourselves into dealing with it. My proposed regex that only cares about the numbers and periods immediately after the ` go` string resolves this as `1.20`, and IMO this is good enough for the purposes of this check.

I changed that testcase and added a couple of others:
```suggestion
assert "1.20" == get_go_version("go version go1.20rc2.10 linux/amd64")
assert "1.22" == get_go_version("go version go1.22-20240109-RC01 cl/597041403 +dcbe772469 X:fieldtrack,boringcrypto linux/amd64")
assert "" == get_go_version("go version goquack-moo linux/amd64")
```

File: site_scons/site_tools/go_builder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mjmac) [reply site_scons/site_tools/go_builder.py:58]: Neat, TIL about pytest.

File: site_scons/site_tools/go_builder.py [hunk 5]
```diff
@@ -63,21 +77,19 @@ def _check_go_version(context):
 
         context.Display(f'Checking {env.d_go_bin} version... ')
         cmd_rc = subprocess.run([env.d_go_bin, 'version'], check=True, stdout=subprocess.PIPE)
-        out = cmd_rc.stdout.decode('utf-8')
+        out = cmd_rc.stdout.decode('utf-8').strip()
         if len(out.split(' ')) < 3:
             context.Result(f'failed to get version from "{out}"')
             return 0
```
Comment (mjmac) [inline site_scons/site_tools/go_builder.py:83]:
```suggestion
```
I would just remove this, as it's not needed anymore with the rest of the changes.

File: site_scons/site_tools/go_builder.py [hunk 6]
```diff
@@ -63,21 +77,19 @@ def _check_go_version(context):
...
 
         # go version go1.2.3 Linux/amd64
-        go_version = out.split(' ')[2].replace('go', '')
-        if '-' in go_version:
-            go_version = go_version.split('-')[0]
+        go_version = get_go_version(out)
```
Comment (mjmac) [inline site_scons/site_tools/go_builder.py:86]:
```suggestion
        go_version = get_go_version(out)
        if go_version == "":
            context.Result(f'failed to get version from "{out}"')
            return 0
```

File: site_scons/site_tools/go_builder.py [hunk 7]
```diff
@@ -63,21 +77,19 @@ def _check_go_version(context):
...
         if len([x for x, y in
                 zip(go_version.split('.'), MIN_GO_VERSION.split('.'))
                 if int(x) < int(y)]) > 0:
-            context.Result(f'{go_version} is too old (min supported: {MIN_GO_VERSION}) ')
+            context.Result(f'{out} is too old (min supported: {MIN_GO_VERSION}) ')
             return 0
-        context.Result(str(go_version))
```
Comment (mjmac) [inline site_scons/site_tools/go_builder.py:80]: Please revert these changes. The scons output looks wacky if the raw output string is included.

**Review by mjmac — CHANGES_REQUESTED** [review]

File: site_scons/site_tools/go_builder.py (hunk 7 above)
Comment (ashleypittman) [reply site_scons/site_tools/go_builder.py:80]: Did you see my other comment, we actually have no idea what the version was that was triggering the failure in the first place because it simply wasn't logged.

Comment (ashleypittman) [reply site_scons/site_tools/go_builder.py:80]: Ah I see now, this is logged if the version can be parsed, it's the other change that's logged if it cannot be.

File: site_scons/site_tools/go_builder.py (hunk 4 above)
Comment (ashleypittman) [reply site_scons/site_tools/go_builder.py]: Done.  I used None rather than "" but the regexp is what you suggested.

**Review by mjmac — APPROVED** [review]
LGTM, thanks for making the updates.

File: site_scons/site_tools/go_builder.py [hunk 8]
```diff
@@ -49,6 +49,26 @@ def _scan_go_file(node, env, _path):
     return includes
 
 
+def get_go_version(output):
+    """Capture only the version after 'go'"""
+    ver_re = re.compile(r'go([0-9\.]+)')
+    groups = ver_re.findall(output)
```
Comment (daltonbohning) [inline site_scons/site_tools/go_builder.py:55]: FYI there usually isn't a need to compile first
```suggestion
    groups = ver_re.findall(r'go([0-9\.]+)', output)
```

**Review by daltonbohning — APPROVED** [review]
